mjpc/planners/sample_gradient/planner.cc

```diff
@@ -0,0 +1,652 @@
+// Copyright 2022 DeepMind Technologies Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mjpc/planners/sample_gradient/planner.h"
+
+#include <absl/random/random.h>
+#include <mujoco/mujoco.h>
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <mutex>
+#include <shared_mutex>
+
+#include "mjpc/array_safety.h"
+#include "mjpc/planners/policy.h"
+#include "mjpc/states/state.h"
+#include "mjpc/trajectory.h"
+#include "mjpc/utilities.h"
+
+namespace mjpc {
+
+namespace mju = ::mujoco::util_mjpc;
+
+// initialize data and settings
+void SampleGradientPlanner::Initialize(mjModel* model, const Task& task) {
+  // delete mjData instances since model might have changed.
+  data_.clear();
+
+  // allocate one mjData for nominal.
+  ResizeMjData(model, 1);
+
+  // model
+  this->model = model;
+
+  // task
+  this->task = &task;
+
+  // rollout parameters
+  timestep_power = 1.0;
+
+  // exploration noise
+  noise_exploration = GetNumberOrDefault(0.1, model, "sampling_exploration");
+
+  // set number of trajectories to rollout
+  num_trajectory_ = GetNumberOrDefault(10, model, "sampling_trajectories");
+
+  // set number of gradient trajectories to rollout
+  num_gradient_ = GetNumberOrDefault(0, model, "gradient_trajectories");
+
+  // gradient filter
+  gradient_filter_ = GetNumberOrDefault(1.0, model, "gradient_filter");
```

Comment:
Did you see any signal from values < 1.0 ?

Reply:
Not really.